src/org/mozilla/javascript/optimizer/BodyCodegen.java

```diff
@@ -3408,26 +3421,54 @@ private static boolean isArithmeticNode(Node node)
             || (type == Token.MUL);
     }
 
-    private void visitArithmetic(Node node, int opCode, Node child,
+    private void visitArithmetic(Node node, int type, Node child,
         Node parent)
     {
         int childNumberFlag = node.getIntProp(Node.ISNUMBER_PROP, -1);
         if (childNumberFlag != -1) {
             generateExpression(child, node);
             generateExpression(child.getNext(), node);
-            cfw.add(opCode);
+
+            switch (type) {
+            case Token.SUB:
+                cfw.add(ByteCode.DSUB);
+                break;
+            case Token.MUL:
+                cfw.add(ByteCode.DMUL);
+                break;
+            case Token.DIV:
+                cfw.add(ByteCode.DDIV);
+                break;
+            case Token.MOD:
+                cfw.add(ByteCode.DREM);
+                break;
+            default:
+                throw Kit.codeBug(Token.typeToName(type));
+            }
         }
         else {
-            boolean childOfArithmetic = isArithmeticNode(parent);
             generateExpression(child, node);
             if (!isArithmeticNode(child))
-                addObjectToDouble();
+                addObjectToNumeric();
             generateExpression(child.getNext(), node);
             if (!isArithmeticNode(child.getNext()))
-                addObjectToDouble();
-            cfw.add(opCode);
-            if (!childOfArithmetic) {
-                addDoubleWrap();
+                addObjectToNumeric();
+
+            switch (type) {
+            case Token.SUB:
+                addScriptRuntimeInvoke("subtract", "(Ljava/lang/Number;Ljava/lang/Number;)Ljava/lang/Number;");
+                break;
+            case Token.MUL:
+                addScriptRuntimeInvoke("multiply", "(Ljava/lang/Number;Ljava/lang/Number;)Ljava/lang/Number;");
```

Comment:
Can we assume that the parameter types are Number objects for these methods? I'm not sure what we do elsewhere but it might be important that we assume that they are Objects instead.

Reply:
The parameter passed here is a Number or BigInt by `addObjectToNumeric`. With Object, it is hard to know explicitly where to do ToNumeric. If you use ToNumeric in `ScriptRuntime.multiply` method, the evaluation order will change, which violates the specification. ref. https://github.com/mozilla/rhino/pull/837#discussion_r579555719